tools/generator/CodeGenerationOptions.cs

```diff
@@ -61,6 +63,47 @@ internal CodeGenerator CreateCodeGenerator (TextWriter writer)
 			}
 		}
 
+		// Encoding format:
+		// - Type name prefix: _JniMarshal_PP
+		// - Parameter types, using JNI encoding, e.g. Z is boolean, I is int, etc. Exception: Reference types, normally encoded as L…;, are instead just L.
+		// - Another _.
+		// - Return type, encoded as with parameters. A void return type is V.
+		internal string GetJniMarshalDelegate (Method method)
+		{
+			var sb = new StringBuilder ("_JniMarshal_PP");
+
+			foreach (var p in method.Parameters)
+				sb.Append (GetJniTypeCode (p.Symbol));
+
+			sb.Append ("_");
+
+			sb.Append (method.IsVoid ? "V" : GetJniTypeCode (method.RetVal.Symbol));
+
+			var result = sb.ToString ();
+
+			lock (jni_marshal_delegates_lock)
+				if (!JniMarshalDelegates.Contains (result))
+					JniMarshalDelegates.Add (result);
+
+			return result;
+		}
+
+		string GetJniTypeCode (ISymbol symbol)
+		{
+			var jni_name = symbol.JniName;
+
+			if (jni_name.StartsWith ("L") || jni_name.StartsWith ("["))
+				return "L";
+
+			return symbol.JniName;
+		}
+
+		internal IEnumerable<string> GetJniMarshalDelegates ()
+		{
+			lock (jni_marshal_delegates_lock)
+				return JniMarshalDelegates.OrderBy (p => p).Distinct ();
```

Comment:
Given that `JniMarshalDelegates` is a `SortedSet<T>`, this `.Distinct()` shouldn't be needed.